src/algebra/category/CommRing/basic.lean

```diff
@@ -18,103 +18,123 @@ We introduce the bundled categories:
 * `CommSemiRing`
 * `CommRing`
 along with the relevant forgetful functors between them.
+
+## Implementation notes
+
+We make SemiRing (and the other categories) locally reducible in order
+to define its instances. This is because writing, for example,
+
+  instance : concrete_category SemiRing := by { delta SemiRing, apply_instance }
+
+results in an instance of the form `id (bundled_hom.concrete_category _)`
+and this `id`, not being [reducible], prevents a later instance search
+(once SemiRing is no longer reducible) from seeing that the morphisms of
+SemiRing are really semiring morphisms (`→+*`), and therefore have a coercion
+to functions, for example. It's especially important that the `has_coe_to_sort`
+instance not contain an extra `id` as we want the `semiring ↥R` instance to
+also apply to `semiring R.α` (it seems to be impractical to guarantee that
+we always access `R.α` through the coercion rather than directly).
+
+TODO: Probably @[derive] should be able to create instances of the
+required form (without `id`), and then we could use that instead of
+this obscure `local attribute [reducible]` method.
+
 -/
 
 universes u v
 
 open category_theory
 
 /-- The category of semirings. -/
-@[reducible] def SemiRing : Type (u+1) := bundled semiring
+def SemiRing : Type (u+1) := bundled semiring
 
 namespace SemiRing
 
 /-- Construct a bundled SemiRing from the underlying type and typeclass. -/
 def of (R : Type u) [semiring R] : SemiRing := bundled.of R
 
+local attribute [reducible] SemiRing
+
+instance : has_coe_to_sort SemiRing := infer_instance
+
 instance (R : SemiRing) : semiring R := R.str
 
 instance bundled_hom : bundled_hom @ring_hom :=
 ⟨@ring_hom.to_fun, @ring_hom.id, @ring_hom.comp, @ring_hom.coe_inj⟩
 
-instance has_forget_to_Mon : has_forget₂ SemiRing.{u} Mon.{u} :=
+instance : concrete_category SemiRing := infer_instance
+
+instance has_forget_to_Mon : has_forget₂ SemiRing Mon :=
```

Comment:
Should we also have a forgetful functor to `AddMon`?

Reply:
Probably to `AddCommMon`. Unfortunately we can't use `bundled_hom.mk_has_forget₂` for this, since AddCommMon is an induced category.

Reply:
I've added this.